### PR TITLE
Add functionality in pipeline to create unmapped bam from fastqs (drop cellranger dependency)

### PR DESCRIPTION
**Pull request by Lioscro**
This PR completely drops cellranger dependency by adding functionality to convert FASTQs to an unmapped BAM and perform (optional) barcode correction to a whitelist.

* Added `preprocess.pipeline.convert_fastqs_to_unmapped_bam`, which converts a set of FASTQs into an unmapped BAM with the appropriate BAM tags.
* Renamed `preprocess.UMI_utils.sort_cellranger_bam` to `sort_bam`.
* Reordered requirements to be alphabetical in `setup.py`.
* Added `ngs-tools>=1.4.0` as a dependency.
* Removed `argparse` from requirements. As of Python 3.2, this is part of the standard library. See https://pypi.org/project/argparse/
* Added (optional) barcode correction `preprocess.pipeline.error_correct_barcodes`. https://kb.10xgenomics.com/hc/en-us/articles/115003822406-How-does-Cell-Ranger-correct-barcode-sequencing-errors
* Remove `_id` argument from `error_correct_umis`. This argument used to be passed to `UMI_utils.correct_umis_in_group`, but it is not used anymore.
* Update command-line pipeline to support new functionality. Renamed `error_correct` to `error_correct_umis` so that it can be differentiated with barcode correction.
* Removed `pandas-charm` from requirements and bumped `pandas>=1.1.4` and `numpy>=1.19.5`

### CODE REVIEW

**Review by mattjones315 — CHANGES_REQUESTED** [review]
Awesome job! The code looks great, and thanks for cleaning up the setup.py a bit! I left one or two comments I'd like to discuss before merging in.

File: cassiopeia/preprocess/pipeline.py [hunk 1]
```diff
@@ -34,6 +35,127 @@
...
+    This function converts a set of FASTQs into an unmapped BAM with appropriate
+    BAM tags.
+
+    Args:
+        fastq_fps: List of paths to FASTQ files. Usually, this argument contains
+            two FASTQs, where the first contains the barcode and UMI sequences
+            and the second contains cDNA. The FASTQs may be gzipped.
+        chemistry: Sample-prep/sequencing chemistry used. The following
+            chemistries are supported:
+            * 10xv3: 10x Genomics 3' version 3; correction performed
```
[the remark this answers is not in the document]
Comment (Lioscro) [reply cassiopeia/preprocess/pipeline.py:60]: No, it wouldn't. I'll add that. Is there anything else that would be helpful? Maybe dropseq?

File: setup.py [hunk 2]
```diff
@@ -13,25 +13,28 @@
...
-    "scikit-bio >= 0.5.6",
+    "matplotlib>=2.2.2",
+    "nbconvert>=5.4.0",
+    "nbformat>=4.4.0",
+    "networkx>=2.5",
+    "ngs-tools>=1.4.0",
+    "numba>=0.51.0",
+    "numpy>1.17",
+    "pandas>=0.22.0",
+    "pandas-charm>=0.1.3",
```
[the remark this answers is not in the document]
Comment (Lioscro) [reply setup.py:30]: Haha. I'll remove it 👍

File: cassiopeia/preprocess/pipeline.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lioscro) [reply cassiopeia/preprocess/pipeline.py]: `10xv2` and `dropseq` added in latest commit

File: cassiopeia/preprocess/pipeline.py [hunk 3]
```diff
@@ -34,6 +35,130 @@
...
+        output_directory, f"{name}_unmapped.bam" if name else "unmapped.bam"
+    )
+    ngs.fastq.fastqs_to_bam_with_chemistry(
+        fastq_fps,
+        ngs.chemistry.get_chemistry(chemistry),
+        tag_map,
+        bam_fp,
+        name=name,
+        show_progress=True,
+        n_threads=n_threads,
```
Comment (mattjones315) [inline cassiopeia/preprocess/pipeline.py:96]: Just making sure - `ngs` appropriately processes the FASTQs according to the chemistry, correct? Specifically, it knows how long cellBCs and UMIs are in the 10X v3 vs 10X v2 chemistry?

File: cassiopeia/preprocess/pipeline.py [hunk 4]
```diff
@@ -34,6 +35,130 @@
...
+    This function converts a set of FASTQs into an unmapped BAM with appropriate
+    BAM tags.
+
+    Args:
+        fastq_fps: List of paths to FASTQ files. Usually, this argument contains
+            two FASTQs, where the first contains the barcode and UMI sequences
+            and the second contains cDNA. The FASTQs may be gzipped.
+        chemistry: Sample-prep/sequencing chemistry used. The following
+            chemistries are supported:
+            * dropseq: Droplet-based scRNA-seq chemistry described in
```
Comment (mattjones315) [inline cassiopeia/preprocess/pipeline.py:60]: Sorry - one more chemistry that would be good to support is inDrop, I think. Does `ngs` know how to deal with this chemistry?

**Review by mattjones315 — COMMENTED** [review]
Nice job! Just left two comments that need not be resolved before merging in. One of them is on supporting inDrop chemistries which would be nice but if it's not easy let's not worry about it.

**Review by mattjones315 — APPROVED** [review]

File: cassiopeia/preprocess/pipeline.py (hunk 3 above)
Comment (Lioscro) [reply cassiopeia/preprocess/pipeline.py:96]: Yes, it does. I also added specific tests for each chemistry.

File: cassiopeia/preprocess/pipeline.py (hunk 4 above)
Comment (Lioscro) [reply cassiopeia/preprocess/pipeline.py:60]: Done

**Review by mattjones315 — APPROVED** [review]
Looks great! Thanks for all the work. I'm going to approve this and merge it in.

**Pull request merged**